north/src/runtime/state.rs

```diff
@@ -407,74 +397,86 @@ impl State {
         Ok(())
     }
 
-    fn is_installed(&self, name: &str, version: &Version) -> bool {
-        match self.applications.get(name) {
-            None => self
-                .resources
-                .get(&(name.to_string(), version.clone())) // TODO get rid of copy
-                .is_some(),
-            Some(app) => app.container.manifest.version == *version,
+    async fn find_installed_npk(
+        &self,
+        name: &str,
+        version: &Version,
+    ) -> Result<Option<(Manifest, PathBuf)>, Error> {
+        // Remove npk from registry
+        for d in &self.config.directories.container_dirs {
+            let mut dir = fs::read_dir(&d)
+                .await
+                .map_err(|e| Error::Io(format!("Failed to read {}", d.display()), e))?;
+            while let Some(res) = dir.next().await {
+                let entry = res.map_err(|e| {
+                    Error::Io(
+                        "Could not read directory".to_string(), // TODO: Which directory?
+                        e,
+                    )
+                })?;
+                let manifest = read_manifest(entry.path().as_path(), &self.signing_keys)
+                    .map_err(Error::Npk)?;
+
+                if manifest.name == name && manifest.version == *version {
+                    return Ok(Some((manifest, entry.path())));
+                }
+            }
         }
+        Ok(None)
     }
 
     /// Remove and umount a specific app
     /// app has to be stopped before it can be uninstalled
     pub async fn uninstall(&mut self, name: &str, version: &Version) -> result::Result<(), Error> {
-        if !self.is_installed(name, version) {
+        let installed_npk = self.find_installed_npk(name, version).await?;
+        if installed_npk.is_none() {
             return Err(Error::ApplicationNotFound);
         }
 
-        let installed_app = self.applications.get(name);
-        let to_uninstall = match installed_app {
-            Some(app) => {
-                if app.is_running() {
-                    warn!("Cannot uninstall started container {}", app);
-                    return Err(Error::ApplicationRunning(app.manifest().name.clone()));
-                }
-                Some(app)
-            }
-            None => self.resources.get(&(name.to_string(), version.clone())),
-        };
-        if let Some(app) = to_uninstall {
+        let (manifest, path) = installed_npk.unwrap();
+
+        let instances = manifest
+            .instances
+            .filter(|i| *i > 1)
+            .map(|i| {
+                (0..i)
+                    .into_iter()
+                    .map(|i| format!("{}-{:03}", manifest.name, i))
+                    .collect()
+            })
+            .unwrap_or_else(|| vec![manifest.name.to_string()]);
+
+        for app in instances.iter().filter_map(|name| {
```

Comment:
That's a little misleading...if there are any running instances, we cannot uninstall the container.
So why not just check if any of the application map keys matches the name + any version and then just return the error?

Reply:
Do you mean that we should do the check for running instances separately before we unmount anything? that makes sense.